### PR TITLE
Add scheduler hook `Addrinfo.getaddrinfo`.

### PR DESCRIPTION
**Pull request by ioquatix**
This hook allows the fiber scheduler to implement asynchronous DNS lookup.

https://bugs.ruby-lang.org/issues/17370

### CODE REVIEW

File: ext/socket/raddrinfo.c [hunk 1]
```diff
@@ -284,8 +284,56 @@ numeric_getaddrinfo(const char *node, const char *service,
...
+    if (res_allocated) { // At least one valid result.
+        return 0;
+    } else {
+        rsock_raise_socket_error("getaddrinfo", EAI_NONAME);
+    }
+}
+
 int
-rb_getaddrinfo(const char *node, const char *service,
+rb_getaddrinfo(VALUE host, const char *node, const char *service,
```
Comment (bruno-) [inline ext/socket/raddrinfo.c:336]: `host` function argument was added so we don't have to allocate new `host` ruby object in `rb_schedule_getaddrinfo` (we just use the argument).

Is this acceptable? If not, we can try other approaches.

File: test/fiber/test_address_resolve.rb [hunk 2]
```diff
@@ -0,0 +1,278 @@
+# frozen_string_literal: true
```
Comment (bruno-) [inline test/fiber/test_address_resolve.rb:1]: I'd like to add a test for every ruby method that touches `rb_schedule_getaddrinfo`.
But this file is already too big so  not sure if that's a good idea.

File: ext/socket/raddrinfo.c (hunk 1 above)
Comment (ioquatix) [reply ext/socket/raddrinfo.c:336]: On the surface of it, it looks like private implementation detail, so it's acceptable. `rb_getaddrinfo` is not exposed in public header right?

File: test/fiber/test_address_resolve.rb (hunk 2 above)
Comment (ioquatix) [reply test/fiber/test_address_resolve.rb:1]: I understand how you feel. I don't have any firm advice about it, except that we want to try and pin down the behaviour as much as "makes sense". Ideally, tests represent the extent of what people would do with the interface, so if we break something, it's picked up by tests before users start complaining :)

File: ext/socket/raddrinfo.c (hunk 1 above)
Comment (bruno-) [reply ext/socket/raddrinfo.c:336]: So, `rb_getaddrinfo` method signature is in `ext/socket/rubysocket.h`. I also made a change there. I'm not sure if that file is a "public header"?

Comment (ioquatix) [reply ext/socket/raddrinfo.c:336]: @nobu is it considered public interface?

**Comment by ioquatix** [comment]
Sorry, I was planning to review this the past weekend but ran out of time. Will defer to the coming weekend.

**Comment by ioquatix** [comment]
Rebased on master.

**Comment by bruno-** [comment]
I worked on changes we talked on slack:

- Add the docs to `doc/fiber.md`
- Refactor `rb_getaddrinfo` function interface - I went a step further in https://github.com/ioquatix/ruby/pull/2

As for the PR, would it be helpful to get on a zoom and do a "guided review" and explain the context?

**Comment by ioquatix** [comment]
I will review this, this weekend.

File: ext/socket/raddrinfo.c (hunk 1 above)
Comment (ioquatix) [reply ext/socket/raddrinfo.c:336]: I found out it's not considered public.

File: ext/socket/raddrinfo.c [hunk 3]
```diff
@@ -297,17 +345,24 @@ rb_getaddrinfo(const char *node, const char *service,
     if (ret == 0)
         allocated_by_malloc = 1;
     else {
+        VALUE scheduler = rb_fiber_scheduler_current();
+
+        if (scheduler != Qnil && node && !(hints->ai_flags & AI_NUMERICHOST) &&
+            strncmp(node, "localhost", strlen(node))) {
```
Comment (ioquatix) [inline ext/socket/raddrinfo.c:351]: There is one interesting concern I have here.

I believe some people will try to use this scheduler hook to do name rewriting. For example, mocking hosts by rewriting DNS.

I wonder if we should therefore avoid making assumptions about the name like `localhost`.

**Comment by ioquatix** [comment]
Rebased on master.

**Comment by ioquatix** [comment]
cc @kirs

**Comment by kirs** [comment]
This is great 👏

**Pull request merged**